### PR TITLE
Trust Graph: implement WASM built-in

### LINKED ISSUE #11: Add add_trust service method with tests

### LINKED ISSUE #16: Add VerificationError in KeyPair::verify instead of SigningError

### PR DESCRIPTION
**Pull request by justprosh**
- Closes #16
- Closes #11 
- build WASM TrustGraph service with peer ids API
- add js example how to use TG: sign trusts/revocations, get certificates
- set up CI
- build TG as a builtin package (like aqua-ipfs or aqua-dht)

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -20,12 +20,16 @@ jobs:
       - run: |
           rustup toolchain install nightly-2021-04-24-x86_64-unknown-linux-gnu
           rustup default nightly-2021-04-24-x86_64-unknown-linux-gnu
+          rustup target add wasm32-wasi --toolchain nightly-2021-04-24-x86_64-unknown-linux-gnu
```
Comment (folex) [inline .circleci/config.yml:23]: Let's move toolchain requirements to `rust-toolchain.toml`? This way, everyone will have the same toolchain.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+          curl -L https://github.com/stedolan/jq/releases/download/jq-1.5/jq-linux64 -o /usr/local/bin/jq
+          chmod +x /usr/local/bin/jq
+
+      - name: Download marine
+        run: bash $GITHUB_WORKSPACE/.github/download_marine.sh
+
+      - uses: actions/setup-java@v2
+        with:
+          distribution: 'adopt'
+          java-version: '11'
```
Comment (folex) [inline .github/workflows/release.yml:36]: I think Java isn't needed here

File: Cargo.toml [hunk 3]
```diff
@@ -1,32 +1,31 @@
 [package]
 name = "trust-graph"
-version = "0.2.8"
+version = "0.2.9"
```
Comment (folex) [inline Cargo.toml:3]: 0.3.0 maybe? You've rewritten the whole thing and changed a lot of APIs. Isn't it a breaking change?

File: admin/README.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# How to generate export certificates
+1. Go to `local-network` 
+2. Run `docker compose up -d` to start Fluence node
+3. Go back to `../admin`
+4. Put `root_secret_key.ed25519` and `issuer_secret_key.ed25519` to folder
```
Comment (folex) [inline admin/README.md:5]: Where do I get these keys?

File: admin/index.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    let root_sk_b58 = fs.readFileSync("./root_secret_key.ed25519").toString();
+    let issuer_sk_b58 = fs.readFileSync("./issuer_secret_key.ed25519").toString();
+    let root_kp = await KeyPair.fromBytes(bs58.decode(root_sk_b58));
+    let issuer_kp = await KeyPair.fromBytes(bs58.decode(issuer_sk_b58));
+    console.log("Root private key: %s", root_sk_b58);
+    console.log("Root peer id: %s", root_kp.Libp2pPeerId.toB58String());
+    console.log("Issuer private key: %s", issuer_sk_b58);
+
+    let node = local[0].peerId;
```
Comment (folex) [inline admin/index.ts:88]: You seem to ignore `environment` here.

File: admin/index.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    fs.writeFileSync("../builtin-package/on_start.json", JSON.stringify({certs: certificates}));
+
+    return;
+}
+
+let environment: Node[];
+environment = local;
+console.log("📘 Will connect to local nodes");
```
Comment (folex) [inline admin/index.ts:113]: Let's add an option to connect to krasnodar, testnet and stage. Here's an example https://github.com/fluencelabs/examples/blob/ff4f4075b92244382416910114dffc6f564d1be4/aqua-examples/aqua-ipfs-integration/nodejs/src/demo.ts#L120-L136

File: admin/package.json [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    "start": "node dist/index.js",
+    "prestart": "npm run build"
+  },
+  "author": "Fluence Labs",
+  "license": "MIT",
+  "dependencies": {
+    "@fluencelabs/aqua": "0.3.1-228",
+    "@fluencelabs/aqua-lib": "0.1.14",
+    "@fluencelabs/fluence": "^0.12.1",
+    "@fluencelabs/fluence-network-environment": "^1.0.10",
```
Comment (folex) [inline admin/package.json:19]: Let's use latest Aqua and FluenceJS

File: aqua/package.json [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    "p2p"
+  ],
+  "author": "Fluence Labs",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/fluencelabs/trust-graph/issues"
+  },
+  "homepage": "https://github.com/fluencelabs/trust-graph#readme",
+  "devDependencies": {
+    "@fluencelabs/aqua": "^0.3.1-228"
```
Comment (folex) [inline aqua/package.json:34]: ditto

File: aqua/trust-graph-api.aqua [hunk 9]
```diff
@@ -0,0 +1,88 @@
+import "trust-graph.aqua"
```
Comment (folex) [inline aqua/trust-graph-api.aqua:1]: Any documentation on all these functions? Link to documentation?

File: aqua/trust-graph-api.aqua [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+                    if cert.chain!1.issued_for == "12D3KooWM45u7AQxsb4MuQJNYT3NWHHMLU7JTbBV66RTfF3KSzdR":
+                        resultBox <<- true
+
+    result_len <- BoolOp.array_length(resultBox)
+    result: *bool
+    if result_len == 0:
+        result <<- false
+    else:
+        result <<- true
+    <- result!
```
Comment (folex) [inline aqua/trust-graph-api.aqua:88]: We decided to move that function inside Rust service, right? If you want to do that in the separate PR, then create an issue and add `TODO:` with link to that issue.

File: builtin-package/on_start.air [hunk 11]
```diff
@@ -0,0 +1,31 @@
+(seq
+    (seq
+        (call relay ("trust-graph" "add_root") ["12D3KooWNbZKaPWRZ8wgjGvrxdJFz9Fq5uVwkR6ERV1f74HhPdyB" 2] add_root_res)
+        (xor
+            (match add_root_res.$.success! true
+                (null)
+           )
+           (call %init_peer_id% ("op" "return") [add_root_res.$.error!])
```
Comment (folex) [inline builtin-package/on_start.air:8]: Since async AquaVM, you need to return to `relay`. Here and everywhere else.

File: admin/index.ts [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    await add_trust_helper(node, root_kp, root_kp.Libp2pPeerId.toB58String(), issuer_kp.Libp2pPeerId.toB58String(), expires_at, cur_time);
+
+    let certificates = [];
+    for (let i = 0; i < krasnodar.length; i++)  {
+        await add_trust_helper(node, issuer_kp, issuer_kp.Libp2pPeerId.toB58String(), krasnodar[i].peerId, expires_at, cur_time);
+        let certs = await get_all_certs(node, krasnodar[i].peerId);
+        certificates.push(certs.certificates[0]);
+    }
+
+    fs.writeFileSync("../builtin-package/on_start.json", JSON.stringify({certs: certificates}));
```
Comment (folex) [inline admin/index.ts:106]: You need to mention that `on_start.json` generation is important. In README, maybe somewhere else as well.

File: .github/workflows/release.yml [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+        id: release
+        uses: softprops/action-gh-release@v1
+        with:
+          name: trust-graph ${{ env.RELEASE_VERSION }}
+          tag_name: ${{ env.RELEASE_VERSION }}
+          files: |
+            trust-graph.tar.gz
+          body: ${{steps.changelog.outputs.changelog}}
+          draft: false
+          prerelease: false
```
Comment (folex) [inline .github/workflows/release.yml:102]:
```suggestion
          prerelease: false
          fail_on_unmatched_files: true
```

File: example/index.ts [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    console.log(
+        "📗 created a fluence peer %s with relay %s",
+        Fluence.getStatus().peerId,
+        Fluence.getStatus().relayPeerId
+    );
+    const issued_timestamp_sec = await timestamp_sec(local[0].peerId);
+    const expires_at_sec = issued_timestamp_sec + 999999999;
+    const issuer_kp = await KeyPair.fromEd25519SK(bs58.decode("29Apzfedhw2Jxh94Jj4rNSmavQ1TkNe8ALYRA7bMegobwp423aLrURxLk32WtXgXHDqoSz7GAT9fQfoMhVd1e5Ww"));
+
+    let add_root_result = await add_root(local[0].peerId, local[0].peerId, 2);
```
Comment (folex) [inline example/index.ts:87]: You're ignoring `environment`

File: example/index.ts [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+    const signed_metadata = await issuer_kp.Libp2pPeerId.privKey.sign(Uint8Array.from(trust_metadata.result));
+
+    let revocation = await issue_revocation(node, revoked_peer_id, revoked_by_peer_id, revoked_at_sec, Array.from(signed_metadata));
+    console.log("Issued revocation %s", revocation.revoke);
+
+    let result_add = await revoke(node, revocation.revoke);
+    console.log("Revoke result: %s", result_add);
+}
+
+async function main(environment: Node[]) {
```
Comment (folex) [inline example/index.ts:73]: This example does poor job of communicating what happens. Plz reconsider what messages you show to a user and how will she learn from these messages.

File: example/package.json [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    "build": "tsc",
+    "start": "node dist/index.js",
+    "prestart": "npm run build"
+  },
+  "author": "Fluence Labs",
+  "license": "MIT",
+  "dependencies": {
+    "@fluencelabs/aqua": "0.4.1-239",
+    "@fluencelabs/aqua-lib": "0.2.0",
+    "@fluencelabs/fluence": "0.14.2",
```
Comment (folex) [inline example/package.json:18]: Different Aqua version from everything else. Update it everywhere to latest, plz.

File: keypair/src/error.rs [hunk 17]
```diff
@@ -58,14 +60,33 @@ pub enum SigningError {
...
         #[source]
-        secp256k1::Error
+        secp256k1::Error,
+    ),
+}
+
+/// An error during verification of a message.
+#[derive(ThisError, Debug)]
+pub enum VerificationError {
+    #[error("Failed to verify signature  with ed25519: {0}")]
```
Comment (folex) [inline keypair/src/error.rs:78]:
```suggestion
    #[error("Failed to verify signature with ed25519: {0}")]
```

File: keypair/src/error.rs [hunk 18]
```diff
@@ -58,14 +60,33 @@ pub enum SigningError {
...
         #[from]
         #[source]
-        secp256k1::Error
+        secp256k1::Error,
+    ),
+}
+
+/// An error during verification of a message.
+#[derive(ThisError, Debug)]
+pub enum VerificationError {
```
Comment (folex) [inline keypair/src/error.rs:77]: I wish the error would show both signature and the public key, but that maybe too much

File: keypair/src/key_pair.rs [hunk 19]
```diff
@@ -217,23 +232,33 @@ impl From<libp2p_core::identity::Keypair> for KeyPair {
...
 
         match key {
-            Ed25519(kp) => Keypair::Ed25519(identity::ed25519::Keypair::decode(kp.encode().to_vec().as_mut_slice()).unwrap()),
+            Ed25519(kp) => Keypair::Ed25519(
+                identity::ed25519::Keypair::decode(kp.encode().to_vec().as_mut_slice()).unwrap(),
+            ),
             #[cfg(not(target_arch = "wasm32"))]
-            Rsa(kp) => Keypair::Rsa(unsafe { std::mem::transmute::<rsa::Keypair, libp2p_core::identity::rsa::Keypair>(kp) }),
-            Secp256k1(kp) => Keypair::Secp256k1(identity::secp256k1::Keypair::from(identity::secp256k1::SecretKey::from_bytes(kp.secret().to_bytes()).unwrap())),
+            Rsa(kp) => Keypair::Rsa(unsafe {
```
Comment (folex) [inline keypair/src/key_pair.rs:256]: `// SAFETY` comment

File: service/Config.toml [hunk 20]
```diff
@@ -0,0 +1,16 @@
+modules_dir = "artifacts/"
```
Comment (folex) [inline service/Config.toml:1]: indentation

File: service/build.rs [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+use marine_rs_sdk_test::generate_marine_test_env;
+use marine_rs_sdk_test::ServiceDescription;
+fn main() {
```
Comment (folex) [inline service/build.rs:18]: nice!

File: service/rust-toolchain [hunk 22]
```diff
@@ -0,0 +1 @@
+nightly
```
Comment (folex) [inline service/rust-toolchain:1]: set specific nightly version

File: service/src/dto.rs [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+    fn try_from(r: Revoke) -> Result<Self, Self::Error> {
+        let revoked_pk = PublicKey::try_from(
+            PeerId::from_str(&r.revoked_peer_id)
+                .map_err(|e| PeerIdDecodeError(format!("{:?}", e)))?,
+        )
+        .map_err(|e| DtoConversionError::PeerIdDecodeError(e.to_string()))?;
+        let revoked_by_pk = PublicKey::try_from(
+            PeerId::from_str(&r.revoked_by).map_err(|e| PeerIdDecodeError(format!("{:?}", e)))?,
+        )
+        .map_err(|e| DtoConversionError::PeerIdDecodeError(e.to_string()))?;
```
Comment (folex) [inline service/src/dto.rs:141]: From the errors, you couldn't tell if error happened when decoding `revoked_peer_id` or `revoked_by`, could you?

To alleviate this, you need to show field name in the error.

File: service/src/dto.rs [hunk 24]
```diff
@@ -0,0 +1,169 @@
...
+        let signature = bs58::decode(&r.signature).into_vec()?;
+        let signature = Signature::from_bytes(KeyFormat::from_str(&r.sig_type)?, signature);
```
Comment (folex) [inline service/src/dto.rs:143]: Here you don't provide any information about which fields caused the error. You sure it's ok?

File: service/src/main.rs [hunk 25]
```diff
@@ -0,0 +1,23 @@
+#![allow(dead_code)]
```
Comment (folex) [inline service/src/main.rs:1]: why?

File: service/src/service_api.rs [hunk 26]
```diff
@@ -0,0 +1,120 @@
...
+    add_root_impl, add_trust_impl, get_all_certs_impl, get_revoke_bytes_impl, get_trust_bytes_impl,
+    get_weight_impl, insert_cert_impl, insert_cert_impl_raw, issue_revocation_impl,
+    issue_trust_impl, revoke_impl, verify_trust_impl, ServiceError,
+};
+use marine_rs_sdk::{marine, CallParameters};
+use trust_graph::MAX_WEIGHT_FACTOR;
+
+#[marine]
+fn get_weight_factor(max_chain_len: u32) -> u32 {
+    MAX_WEIGHT_FACTOR - max_chain_len
```
Comment (folex) [inline service/src/service_api.rs:17]: Underflow check?

File: service/src/service_api.rs [hunk 27]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+#[marine]
+/// could add only a owner of a trust graph service
+fn add_root(peer_id: String, weight_factor: u32) -> AddRootResult {
+    let call_parameters: CallParameters = marine_rs_sdk::get_call_parameters();
+    let init_peer_id = call_parameters.init_peer_id.clone();
```
Comment (folex) [inline service/src/service_api.rs:24]: Why clone?

File: service/src/service_api.rs [hunk 28]
```diff
@@ -0,0 +1,120 @@
+use crate::dto::{Certificate, Revoke, Trust};
```
Comment (folex) [inline service/src/service_api.rs:1]: Docs on public methods?

**Review by folex — CHANGES_REQUESTED** [review]

File: admin/README.md (hunk 4 above)
Comment (justprosh) [reply admin/README.md:5]: i will store in some place in notion, I think. You can suggest a better alternative

**Review by folex — APPROVED** [review]

**Pull request merged**